rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,498 @@
+  /** 
+   * Get a Background Session object.
+   * @param request 
+   * @param sessionId
+   * @returns the Background Session object if exists, or undefined.
+   */
+  get: async (
+    request: KibanaRequest,
+    sessionId: string
+  ) => Promise<BackgroundSessionAttributes?>
+
+  /** 
+   * Get a searchId from a Background Session object.
+   * @param request 
+   * @param sessionId
+   * @returns the searchID if exists on the Background Session, or undefined.
+   */
+  getSearchId: async (
+    request: KibanaRequest,
+    sessionId: string
+  ) => Promise<string?>
+
+   /**
+    * Store a session.
+    * @param request 
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @param searchIdMap A mapping of hashed requests mapped to the corresponding searchId.
+    * @param url TODO: is the URL provided here? How?
+    * @returns The stored `BackgroundSessionAttributes` object
+    * @throws Throws an error in OSS.
+    * @internal (Consumers should use searchInterceptor.sendToBackground())
+    */
+  store: (
+    request: KibanaRequest,
+    sessionId: string,
+    name: string,
+    url: string,
+    searchIdMapping?: Record<string, string>
+  ) => Promise<BackgroundSessionAttributes>
+
+   /**
+    * Mark a session as and all associated searchIds as expired.
+    * @param request 
+    * @param sessionId
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  expire: async (
+    request: KibanaRequest,
+    sessionId: string
+  ) => Promise<boolean>
+
+   /**
+    * Extend a session and all associated searchIds.
+    * @param request 
+    * @param sessionId 
+    * @param extendBy Time to extend by, can be a relative or absolute string.
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  extend: async (
+    request: KibanaRequest,
+    sessionId: string,
+    extendBy: string,
+  ) => Promise<boolean>
+
+   /**
+    * Get a list of background session objects.
+    * @param request 
+    * @param sessionId
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  list: async (
+    request: KibanaRequest,
+  ) => Promise<BackgroundSessionAttributes[]>
+
+   /**
+    * Update the status of a given session
+    * @param request 
+    * @param sessionId
+    * @param status
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  updateStatus: async (
+    request: KibanaRequest,
+    sessionId: string,
+    status: BackgroundSessionStatus
+  ) => Promise<boolean>
+}
+
+```
+
+### Search Service Changes
+
+There are cases where search requests are issued by the server (Like TSVB).
+We can simplify this flow by introducing a mechanism, similar to the frontend one, tracking the information in memory. 
+
+However, due to the fact that we are running in a potentially multi-server architecture, each server will have to poll for a saved object with a corresponding sessionId to store the ids into it.
```

Comment:
i don't understand this, can it be further explained ?
we have multi server architecture even when using this on the client, and we are polling inside server side search service for the session saved object. how is this different ?

also, when using tsvb, session is still initiated from the client (and will be stored from the client) so how does tsvb make this different ?

Reply:
There are two mechanisms at play here:

1. When all search requests are initiated by the client, the search IDs can be passed back and the request/search ID map can be built client-side, and sent in the request to "send to background". No polling for a saved object is needed in this case.
1. When some search requests are initiated by the server. In this case, either we expect the server to send the search ID back to the client so the client-side request/search ID mapping can be updated, or alternatively, the server can poll for the saved object and update the object with the search ID it got back (without needing to send the ID back to the client).